interfaces/builtin/microstack_support.go

```diff
@@ -0,0 +1,247 @@
+
+/*
+ * Microstack is a full OpenStack in a single snap package.
+ * Virtual machines are spawned as QEMU processes with libvirt acting as a management
+ * daemon (including for activities such as applying AppArmor profiles).
+ * Networking is provided largely via OpenVSwitch and Neutron with dnsmasq acting
+ * as an auxiliary daemon. tun/tap kernel module is used for creating virtual interfaces.
+ * Virtual machines rely on KVM for virtualization acceleration and on vhost
+ * framework in the kernel (vhost_net, vhost_scsi, vhost_vsock).
+ */
+
+const microStackSupportSummary = `allows operating as the MicroStack service`
+
+const microStackSupportBaseDeclarationPlugs = `
+  microstack-support:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const microStackSupportBaseDeclarationSlots = `
+  microstack-support:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const microStackSupportConnectedPlugAppArmor = `
+
+# Used by QEMU to work with the kernel-side virtio implementations.
+/dev/vhost-net rw,
+/dev/vhost-scsi rw,
+/dev/vhost-vsock rw,
+# Used by QEMU to work with VFIO (https://www.kernel.org/doc/Documentation/vfio.txt).
+# For vfio hotplug on systems without static vfio (LP: #1775777)
+# VFIO userspace driver interface.
+/dev/vfio/vfio rw,
+# Access to VFIO group character devices such as /dev/vfio/<group> where <group> is the group number.
+/dev/vfio/* rw,
+# Used by Nova for mounting images via qemu-nbd.
+/dev/nbd* rw,
+
+# Allow creating dm-* devices, /dev/<vg-name> directories, /dev/mapper directory and symlinks under it.
+# Allow issuing ioctls to the Device Mapper for LVM tools via /dev/mapper/control.
+/dev/** rw,
+# Allow access to loop devices and loop-control to be able to associate a file with a loop device
+# for the purpose of using a file-backed LVM setup.
+/dev/loop-control rw,
+/dev/loop[0-9]* rw,
+
+# Description: this policy intentionally allows Microstack services to configure AppArmor
+# as libvirt generates AppArmor profiles for the utility processes it spawns.
+/sys/kernel/security/apparmor/{,**} r,
+/sys/kernel/security/apparmor/.remove w,
+/sys/kernel/security/apparmor/.replace w,
+
+# Used by libvirt to work with IOMMU.
+/sys/kernel/iommu_groups/{,**} r,
+/sys/bus/pci/devices/**/iommu_group/** r,
+
+# Used by libvirt's QEMU driver state initialization code path.
+# The path used is hard-coded in libvirt to <huge-page-mnt-dir>/libvirt/qemu.
+/dev/hugepages/libvirt/ rw,
+/dev/hugepages/libvirt/** mrwklix,
+
+# Used by QEMU to get the maximum number of memory regions allowed in the vhost kernel module.
+/sys/module/vhost/parameters/max_mem_regions r,
+
+# Used by libvirt (cgroup-related):
+/sys/fs/cgroup/unified/cgroup.controllers r,
+/sys/fs/cgroup/cpuset/cpuset.cpus r,
+
+# Non-systemd layout: https://libvirt.org/cgroups.html#currentLayoutGeneric
+/sys/fs/cgroup/*/ r,
+/sys/fs/cgroup/*/machine/ rw,
+/sys/fs/cgroup/*/machine/** rw,
```

Comment:
This was discussed above earlier (see https://github.com/snapcore/snapd/pull/8926#discussion_r449584383 for an in-depth explanation with code references) - it has to do with presence of `systemd-machined` that comes with the `systemd-container` on Ubuntu (not present in our cloud images by default).

Since `systemd-machined` isn't present, the generic layout is used on a system with systemd present.

Hence those rules are here.

Reply:
Ah thanks for the pointer, I think I had read that comment originally a few weeks/months ago but forgot about it when I did another review again